### PR TITLE
Refactor and speed up /api/action/people (trends person modal)

### PR DESCRIPTION
**Pull request by macobo**
## Changes

Goals:
1. Speed up calculating trends people
1. Simplify the code as much as possible
1. Use similar patterns as other code WRT querying

How:
1. Using joins instead of subqueries for person properties speeds up the queries a lot. In testing this sped up a sample query 10x (10s -> 1s). [Slack thread](https://posthog.slack.com/archives/C02283301EK/p1631009988030800)
2. Use the relatively recent TrendsEventQuery to avoid needing to reimplement all the logic inside trends in this query
3. Separate class for the query over the endpoint

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

File: ee/clickhouse/views/test/test_clickhouse_action.py [hunk 1]
```diff
@@ -257,3 +257,83 @@ def test_breakdown_by_event_property_none_people_endpoint(self):
             },
         ).json()
         self.assertEqual(len(people["results"][0]["people"]), 1)
+
+    def _test_interval(self, date_from, interval, timestamps):
```
Comment (macobo) [inline ee/clickhouse/views/test/test_clickhouse_action.py:261]: Note (cc @EDsCODE) - these tests mostly all fail on postgres since we don't do any special-case date handling there. Thoughts, shall I just file a bug in the backlog (want to avoid scope creep)?

File: ee/clickhouse/queries/event_query.py [hunk 2]
```diff
@@ -41,6 +46,8 @@ def __init__(
 
         self._should_join_distinct_ids = should_join_distinct_ids
         self._should_join_persons = should_join_persons
+        self._extra_fields = extra_fields
+        self._extra_person_fields = extra_person_fields
```
Comment (hazzadous) [inline ee/clickhouse/queries/event_query.py:50]: I see that `self._extra_person_fields` is being used within the class but I don't see `self._extra_fields` being used, do we need it?

Comment (hazzadous) [reply ee/clickhouse/queries/event_query.py:50]: Also, what does extra_person_fields mean? Is it just that these fields will be selected? Perhaps a comment to say so would be useful

Comment (macobo) [reply ee/clickhouse/queries/event_query.py:50]: `_extra_fields` is being used in a subclass when constructing the fields to select from (events) table :)

Will add a comment

File: ee/clickhouse/queries/person_query.py [hunk 3]
```diff
@@ -1,36 +1,49 @@
...
                 if self._column_optimizer.should_query_person_properties_column
+                or self.PERSON_PROPERTIES_ALIAS in self._extra_fields
                 else ""
             )
             + " ".join(
                 f", argMax({column_name}, _timestamp) as {column_name}"
                 for column_name in self._column_optimizer.materialized_person_columns_to_query
             )
+            + " ".join(
+                f", argMax({column_name}, _timestamp) as {self.ALIASES.get(column_name, column_name)}"
```
Comment (hazzadous) [inline ee/clickhouse/queries/person_query.py:44]: I gather the argMax is simulating something like a DISTINCT ON(id) query? Is there any reason we don't use this instead?

Comment (macobo) [reply ee/clickhouse/queries/person_query.py:44]: https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/reference/argmax/

Since there can be multiple rows with the same id (due to updates), we need to select latest properties. DISTINCT would not help us beyond getting the id.

> Calculates the arg value for a maximum val value. If there are several different values of arg for maximum values of val, returns the first of these values encountered.

This is what we need here to get the "latest" column value for this person from CH.

File: ee/clickhouse/queries/trends/person.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+from posthog.models.team import Team
+
+
+def _handle_date_interval(filter: Filter) -> Filter:
+    # adhoc date handling. parsed differently with django orm
+    date_from = filter.date_from or timezone.now()
+    data: Dict = {}
+    if filter.interval == "month":
+        data.update(
+            {"date_to": (date_from + relativedelta(months=1) - timedelta(days=1)).strftime("%Y-%m-%d %H:%M:%S")}
```
Comment (hazzadous) [inline ee/clickhouse/queries/trends/person.py:26]: I gather from this that `date_from` and `date_to` are inclusive, hence the timedelta days=1?  What would the filter be doing in this case? Would we want to include all of "date_to" and "date_from"? and should we be setting the time component to midnight?

Out of scope but I wonder if it would make sense to use a half open interval [date_from, date_to) here instead when doing date filtering so we don't have to think about details like this.

Comment (hazzadous) [reply ee/clickhouse/queries/trends/person.py:26]: Ah I see this is just a copy of previous code, ignore as out of scope.

File: ee/clickhouse/queries/trends/trend_event_query.py [hunk 5]
```diff
@@ -31,6 +31,17 @@ def get_query(self) -> Tuple[str, Dict[str, Any]]:
                 )
             )
             + (f", {self.DISTINCT_ID_TABLE_ALIAS}.person_id as person_id" if self._should_join_distinct_ids else "")
+            + (
+                " ".join(
+                    f", {self.EVENT_TABLE_ALIAS}.{column_name} as {column_name}" for column_name in self._extra_fields
```
Comment (hazzadous) [inline ee/clickhouse/queries/trends/trend_event_query.py:36]: I gather we don't need to do argMax here because there will only be one vary for each event_id, column?

File: ee/clickhouse/queries/trends/person.py (hunk 4 above)
Comment (macobo) [reply ee/clickhouse/queries/trends/person.py:26]: This is one of those gnarly bits here, see also my comment to Eric. Newer code handles date ranges properly without endpoint-specific wrangling.

Your assumption for date range is kind of incorrect. Fundamentally all clickhouse date filters (assuming interval allows) end up as `timestamp >= '2021-09-10 00:00:00 AND timestamp <= '2021-09-10 23:59:59'`, but the end-of-date logic is handled separately here.

File: ee/clickhouse/queries/trends/trend_event_query.py (hunk 5 above)
Comment (macobo) [reply ee/clickhouse/queries/trends/trend_event_query.py:36]: We don't do a group by here. In persons table we need to select which value to use due to group by :)

**Comment by macobo** [comment]
ping @hazzadous @EDsCODE would love to get this approved, let me know what are the blocking issues (besides the missing comment) :)

**Comment by hazzadous** [comment]
Nothing from me @macobo although my approval isn’t so inform yet on the click house querying side

File: ee/clickhouse/queries/person_query.py [hunk 6]
```diff
@@ -1,36 +1,49 @@
...
+from typing import List, Optional, Union
 
+from ee.clickhouse.materialized_columns.columns import ColumnName
 from ee.clickhouse.queries.column_optimizer import ColumnOptimizer
 from posthog.models import Filter
 from posthog.models.filters.path_filter import PathFilter
 
 
 class ClickhousePersonQuery:
     PERSON_PROPERTIES_ALIAS = "person_props"
```
Comment (EDsCODE) [inline ee/clickhouse/queries/person_query.py:10]: Delete this now and just use the dict in the future?

File: ee/clickhouse/sql/person.py [hunk 7]
```diff
@@ -320,3 +299,22 @@
...
 """
+
+GET_PERSONS_FROM_EVENT_QUERY = """
+SELECT
+    person_id,
+    created_at,
+    team_id,
+    person_props,
+    is_identified,
+    arrayReduce('groupUniqArray', groupArray(distinct_id)) AS distinct_ids
```
Comment (EDsCODE) [inline ee/clickhouse/sql/person.py:310]: still worried this could be a mess with some of person_ids with uniquely large amount of distinct_ids 🤷‍♂️

**Review by EDsCODE — APPROVED** [review]
Very satisfying refactor both in better code reusability and readability

I'm taking your word that the new query is more performant because it follows the same patterns we improved on elsewhere in combining the person_props filtering

File: ee/clickhouse/queries/person_query.py (hunk 6 above)
Comment (macobo) [reply ee/clickhouse/queries/person_query.py:10]: Other places rely on this externally

File: ee/clickhouse/sql/person.py (hunk 7 above)
Comment (macobo) [reply ee/clickhouse/sql/person.py:310]: Agreed - we should look into making person links person_id based at some point which would solve this mess. We _could_ also just do `[any(distinct_id)] AS distinct_ids` but :man_shrugging:
